File: communicationAPI/src/test/java/eu/cybergeiger/communication/TestSerialization.java
package eu.cybergeiger.communication;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.UUID;
import org.junit.Test;

/**
 * <p>Unit testing of serializable objects.</p>
 */
public class TestSerialization {

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  @Test
  public void parameterListSerializationTest() throws IOException {
    ParameterList p = new ParameterList("test", null, "Test", "null");

    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    ParameterList p2 = ParameterList.fromByteArrayStream(bin);
    assertEquals("Cloned ParameterLists are not equal", p.toString(), p2.toString());

  }

  /**
   * <p>Tests the serialization of the GeigerUrl object.</p>
   */
  @Test
  public void geigerUrlSerializationTest() throws IOException {
    GeigerUrl p = new GeigerUrl("id", "path");
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    GeigerUrl p2 = GeigerUrl.fromByteArrayStream(bin);
    assertEquals("Cloned MenuItems are not equal", p.toString(), p2.toString());

  }

  /**
   * <p>Tests the serialization of the MenuItem object.</p>
   */
  @Test
  public void menuItemSerializationTest() throws IOException {
    MenuItem p = new MenuItem("test", new GeigerUrl("Test", "action"), false);

    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    MenuItem p2 = MenuItem.fromByteArrayStream(bin);
    assertEquals("Cloned MenuItems are not equal", p.toString(), p2.toString());

  }

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  @Test
  public void pluginInformationSerializationTest() throws IOException {
    PluginInformation p = new PluginInformation("exec", 1234);
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    PluginInformation p2 = PluginInformation.fromByteArrayStream(bin);
    PluginInformation p3 = PluginInformation.fromByteArray(p.toByteArray());
    assertNotNull(p);
    assertNotNull("deserialized object (stream) is null", p2);
    assertNotNull("deserialized object (array) is null", p3);
    assertEquals("Cloned Plugininformation using stream are not equal",
        p.hashCode(), p2.hashCode());
    assertEquals("Cloned Plugininformation using array are not equal",
        p.hashCode(), p3.hashCode());

  }

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  @Test
  public void communicationSecretSerializationTest() throws IOException {
    CommunicationSecret[] comsec = new CommunicationSecret[]{
        new CommunicationSecret("Hello".getBytes(StandardCharsets.UTF_8)),
        new CommunicationSecret(),
        new CommunicationSecret(1024)
      };

    for (CommunicationSecret p : comsec) {
      ByteArrayOutputStream bout = new ByteArrayOutputStream();
      p.toByteArrayStream(bout);
      ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
      CommunicationSecret p2 = CommunicationSecret.fromByteArrayStream(bin);
      assertNotNull(p);
      assertNotNull("deserialized object (stream) is null", p2);
      assertEquals("Cloned Plugininformation using stream are not equal",
          p.toString(), p2.toString());
    }
  }

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  @Test
  public void storableMapSerializationTest() throws IOException {
    StorableHashMap hm = new StorableHashMap();
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    hm.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    StorableHashMap hm2 = new StorableHashMap();
    StorableHashMap.fromByteArrayStream(bin, hm2);
    assertNotNull("deserialized object (stream) is null", hm2);
    assertEquals("Cloned Plugininformation using stream are not equal",
        hm.toString(), hm2.toString());
  }

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  @Test
  public void messageSerializationTest() throws IOException {
    for (Message m :
        new Message[]{
            new Message("src", "target", MessageType.DEACTIVATE_PLUGIN,
                new GeigerUrl("geiger://id1/path1"), null),
            new Message("src", "target", MessageType.DEACTIVATE_PLUGIN,
                null, new byte[0]),
            new Message("src", "target", MessageType.DEACTIVATE_PLUGIN,
                new GeigerUrl("id", "path2"), new byte[0]),
            new Message("src", "target", MessageType.DEACTIVATE_PLUGIN,
                new GeigerUrl("id", null), new byte[0]),
            new Message("src", "target", MessageType.DEACTIVATE_PLUGIN,
                new GeigerUrl("id1", null), new byte[0]),
            new Message("src", "target", null, null, new byte[0]),
            new Message("src", "target", null, null,
                UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8))
          }
    ) {
      ByteArrayOutputStream bout = new ByteArrayOutputStream();
      m.toByteArrayStream(bout);
      ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
      Message m2 = Message.fromByteArray(bin);
      assertNotNull("deserialized object (stream) is null", m2);
      assertEquals("Cloned Plugininformation using stream are not equal", m, m2);
    }
  }

}

File: test/test_serialization.dart
library geiger_api;

import 'dart:async';

import 'package:geiger_api/geiger_api.dart';
import 'package:geiger_api/src/communication/communication_secret.dart';
import 'package:geiger_api/src/communication/geiger_url.dart';
import 'package:geiger_api/src/communication/menu_item.dart';
import 'package:geiger_api/src/communication/parameter_list.dart';
import 'package:geiger_api/src/communication/plugin_information.dart';
import 'package:geiger_localstorage/geiger_localstorage.dart';
import 'package:test/expect.dart';
import 'package:test/scaffolding.dart';
import 'package:uuid/uuid.dart';

class TestSerialization {}

/// Unit testing of serializable objects.
void main() {
  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  test('parameterListSerializationTest', () async {
    ParameterList p = ParameterList(['test', '', 'Test', 'null']);
    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream tempStream =
        ByteStream(Stream<List<int>>.value(await bout.bytes));
    ParameterList? p2 = await ParameterList.fromByteArrayStream(tempStream);
    expect(p.toString() == p2.toString(), true,
        reason: 'Cloned Parameter lists are not equal');
  });

  /**
   * <p>Tests the serialization of the GeigerUrl object.</p>
   */
  test('geigerUrlSerializationTest', () async {
    GeigerUrl p = GeigerUrl(null, GeigerApi.masterId, 'path');
    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));

    /// TODO(unassigned): What is that below?
    await GeigerUrl.fromByteArrayStream(bin);

    expect(p.toString() == p.toString(), true,
        reason: 'Cloned GeigerUrls are unequal');
  });

  /**
   * <p>Tests the serialization of the MenuItem object.</p>
   */
  test('menuItemSerializationTest', () async {
    MenuItem p =
        MenuItem('test', GeigerUrl(null, GeigerApi.masterId, 'path'), false);

    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(null, await bout.bytes);
    MenuItem p2 = await MenuItem.fromByteArrayStream(bin);
    expect(p.toString() == p2.toString(), true,
        reason: 'Cloned MenuItems are unequal');
  });

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  test('pluginInformationSerializationTest', () async {
    PluginInformation p =
        PluginInformation('exec', 1234, CommunicationSecret('Hello'.codeUnits));
    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));
    PluginInformation p2 = await PluginInformation.fromByteArrayStream(bin);
    PluginInformation p3 =
        await PluginInformation.fromByteArray(await p.toByteArray()!);
    expect(p.hashCode == p2.hashCode, true,
        reason: 'Cloned Plugininformation using stream are not equal');
    expect(p.hashCode == p3.hashCode, true,
        reason: 'Cloned Plugininformation using array are not equal');
  });

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  test('communcationSecretSerialization', () async {
    List<CommunicationSecret> comsec = [
      CommunicationSecret('hello'.codeUnits),
      CommunicationSecret('a'.codeUnits),
      CommunicationSecret([35])
    ];

    for (CommunicationSecret p in comsec) {
      ByteSink bout = ByteSink();
      p.toByteArrayStream(bout);
      bout.close();
      ByteStream bin = ByteStream(null, await bout.bytes);
      CommunicationSecret p2 =
          await CommunicationSecret.fromByteArrayStream(bin);
      expect(p.toString() == p2.toString(), true,
          reason: 'Cloned Plugininformation using stream are not equal');
    }
  });

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  test('storableMapSerializationTest', () async {
    StorableHashMap<GeigerUrl, GeigerUrl> hm =
        StorableHashMap<GeigerUrl, GeigerUrl>();
    Map<GeigerUrl, GeigerUrl> it = <GeigerUrl, GeigerUrl>{
      GeigerUrl(null, GeigerApi.masterId, 'path'):
          GeigerUrl(null, GeigerApi.masterId, 'path')
    };
    hm.addAll(it);
    ByteSink bout = ByteSink();
    hm.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));
    StorableHashMap hm2 = StorableHashMap();
    await StorableHashMap.fromByteArrayStream(bin, hm2);
    expect(hm2.toString() == hm.toString(), true,
        reason: 'Cloned Plugininformation using stream are not equal');
  });

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  test('messageSerializationTest', () async {
    List<Message> messageList = [
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, 'geiger://id1/path1'), null),
      Message('src', 'target', MessageType.deactivatePlugin, null, []),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, 'path2'), []),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, ''), <int>[]),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, ''), <int>[]),
      Message('src', 'target', MessageType.allEvents, null, <int>[]),
      Message('src', 'target', MessageType.allEvents, null,
          const Uuid().v4().codeUnits)
    ];

    for (Message m in messageList) {
      ByteSink bout = ByteSink();
      m.toByteArrayStream(bout);
      bout.close();
      ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));
      Message m2 = await Message.fromByteArray(bin);
      expect(m == m2, true,
          reason: 'Cloned Plugininformation using stream are not equal');
    }
  });
}
